<a href="https://colab.research.google.com/github/o001705/IIScProjects/blob/main/M1_AST_01_Linear_Algebra_A.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Advanced Programme in Deep Learning (Foundations and Applications)
## A Program by IISc and TalentSprint
### Assignment 1: Linear Algebra

## Learning Objectives

At the end of the assignment, you will be able to
* understand matrix, types of matrix, vectors and its related operations
* solve the system of linear equation using matrix solution
* implement concepts related to eigen vectors and values



### Setup Steps:

In [2]:
#@title Please enter your registration id to start: { run: "auto", display-mode: "form" }
Id = "2237536" #@param {type:"string"}

In [3]:
#@title Please enter your password (normally your phone number) to continue: { run: "auto", display-mode: "form" }
password = "9892920569" #@param {type:"string"}

In [4]:
#@title Run this cell to complete the setup for this Notebook
from IPython import get_ipython
import warnings
warnings.filterwarnings("ignore")

ipython = get_ipython()
  
notebook= "M1_AST_01_Linear_Algebra_A" #name of the notebook

def setup():
    from IPython.display import HTML, display
    display(HTML('<script src="https://dashboard.talentsprint.com/aiml/record_ip.html?traineeId={0}&recordId={1}"></script>'.format(getId(),submission_id)))
    print("Setup completed successfully")
    return

def submit_notebook():
    ipython.magic("notebook -e "+ notebook + ".ipynb")
    
    import requests, json, base64, datetime

    url = "https://dashboard.talentsprint.com/xp/app/save_notebook_attempts"
    if not submission_id:
      data = {"id" : getId(), "notebook" : notebook, "mobile" : getPassword()}
      r = requests.post(url, data = data)
      r = json.loads(r.text)

      if r["status"] == "Success":
          return r["record_id"]
      elif "err" in r:        
        print(r["err"])
        return None        
      else:
        print ("Something is wrong, the notebook will not be submitted for grading")
        return None
    
    elif getAnswer1() and getAnswer2() and getComplexity() and getAdditional() and getConcepts() and getComments() and getMentorSupport():
      f = open(notebook + ".ipynb", "rb")
      file_hash = base64.b64encode(f.read())

      data = {"complexity" : Complexity, "additional" :Additional, 
              "concepts" : Concepts, "record_id" : submission_id, 
              "answer1" : Answer1, "answer2" : Answer2, "id" : Id, "file_hash" : file_hash,
              "notebook" : notebook,
              "feedback_experiments_input" : Comments,
              "feedback_mentor_support": Mentor_support}
      r = requests.post(url, data = data)
      r = json.loads(r.text)
      if "err" in r:        
        print(r["err"])
        return None   
      else:
        print("Your submission is successful.")
        print("Ref Id:", submission_id)
        print("Date of submission: ", r["date"])
        print("Time of submission: ", r["time"])
        print("View your submissions: https://dlfa-iisc.talentsprint.com/notebook_submissions")
        #print("For any queries/discrepancies, please connect with mentors through the chat icon in LMS dashboard.")
        return submission_id
    else: submission_id
    

def getAdditional():
  try:
    if not Additional: 
      raise NameError
    else:
      return Additional  
  except NameError:
    print ("Please answer Additional Question")
    return None

def getComplexity():
  try:
    if not Complexity:
      raise NameError
    else:
      return Complexity
  except NameError:
    print ("Please answer Complexity Question")
    return None
  
def getConcepts():
  try:
    if not Concepts:
      raise NameError
    else:
      return Concepts
  except NameError:
    print ("Please answer Concepts Question")
    return None
  
  
# def getWalkthrough():
#   try:
#     if not Walkthrough:
#       raise NameError
#     else:
#       return Walkthrough
#   except NameError:
#     print ("Please answer Walkthrough Question")
#     return None
  
def getComments():
  try:
    if not Comments:
      raise NameError
    else:
      return Comments
  except NameError:
    print ("Please answer Comments Question")
    return None
  

def getMentorSupport():
  try:
    if not Mentor_support:
      raise NameError
    else:
      return Mentor_support
  except NameError:
    print ("Please answer Mentor support Question")
    return None

def getAnswer1():
  try:
    if not Answer1:
      raise NameError 
    else: 
      return Answer1
  except NameError:
    print ("Please answer Question 1")
    return None

def getAnswer2():
  try:
    if not Answer2:
      raise NameError 
    else: 
      return Answer2
  except NameError:
    print ("Please answer Question 2")
    return None
  

def getId():
  try: 
    return Id if Id else None
  except NameError:
    return None

def getPassword():
  try:
    return password if password else None
  except NameError:
    return None

submission_id = None
### Setup 
if getPassword() and getId():
  submission_id = submit_notebook()
  if submission_id:
    setup() 
else:
  print ("Please complete Id and Password cells before running setup")



Setup completed successfully


### Importing required packages

In [5]:
import numpy as np
from math import *
from numpy import array
from numpy import diag
from numpy import dot
from numpy import zeros
from numpy import linalg as la
from numpy.linalg import eig

### Types of Matrix

**Square Matrix**: In square matrix, the number of rows are equal to the number of columns.

1. Create a square matrix with elements 61 to 69. 

In [6]:
elements = np.arange(61, 70)
matrix = elements.reshape(3,3) 
print(matrix)

[[61 62 63]
 [64 65 66]
 [67 68 69]]



**Upper Triangular Matrix**: It is a type of square matrix, where all the values below the main diagonal are 0.

**Lower Triangular Matrix**: It is a type of square matrix, where all the values above the main diagonal are 0.


2. Create an upper and lower triangular matrix of size 6x6 with random values.

In [8]:
matrix = np.random.rand(6,6)
print("Original Matrix:\n",matrix)

print("\n\n")
# Upper triangular matrix
upper = np.triu(np.random.rand(6,6))

# Lower triangular matrix
lower = np.tril(np.random.rand(6,6))

print("Upper Triangular Matrix: \n", upper)
print("\n\n")
print("Lower Triangular Matrix: \n", lower)


Original Matrix:
 [[0.6314381  0.65475009 0.643245   0.41754969 0.03873285 0.55504902]
 [0.23336809 0.49405638 0.34601621 0.65753    0.97028147 0.23532847]
 [0.20088666 0.34156654 0.87965854 0.65594495 0.24455747 0.81755093]
 [0.68987462 0.6133764  0.51203911 0.0319479  0.31977782 0.37459817]
 [0.30270395 0.07799914 0.83626332 0.06488526 0.3136098  0.44924014]
 [0.64291772 0.09156858 0.64094699 0.25371708 0.03534815 0.83444893]]



Upper Triangular Matrix: 
 [[0.86074377 0.42131026 0.75956416 0.46486212 0.30637152 0.32733192]
 [0.         0.11000644 0.01479844 0.75976963 0.86495232 0.63216822]
 [0.         0.         0.81799008 0.45886487 0.784527   0.32158289]
 [0.         0.         0.         0.33064719 0.38489243 0.82352097]
 [0.         0.         0.         0.         0.92775814 0.98121323]
 [0.         0.         0.         0.         0.         0.98078132]]



Lower Triangular Matrix: 
 [[0.13023795 0.         0.         0.         0.         0.        ]
 [0.32031945 0.0210647 

**Diagonal Matrix**: In a diagonal matrix, all the values are zero except the diagonal values

3. Create a diagonal matrix of size 3x3 with integer valules and do the following:

  i). Create a diagonal matrix with the values above the diagonal elements.
  
  ii). Create a diagonal matrix with the values below the diagonal elements.





In [15]:
M = np.array([np.arange(73,76),np.arange(32,35), np.arange(57,60)])
print("Matrix:\n",M)
 
# Diagonal Matrix
diagonal_values = np.array([73, 33, 59])
DM = np.diag(diagonal_values)
print("\n\nDiagonal Matrix:\n",DM)


# Using k=1 for getting the diagonal matrix with the values above the diagonal elements
ADV = np.array([74, 34])
ADM = np.diag(ADV, k=1)
print("\n\nAbove Diagonal Matrix:\n",ADM)

# Using k=-1 for getting the diagonal matrix with the values below the diagonal elements
BDV = np.array([32, 58])
BDM = np.diag(BDV, k=-1)
print("\n\nBelow Diagonal Matrix:\n",BDM)


Matrix:
 [[73 74 75]
 [32 33 34]
 [57 58 59]]


Diagonal Matrix:
 [[73  0  0]
 [ 0 33  0]
 [ 0  0 59]]


Above Diagonal Matrix:
 [[ 0 74  0]
 [ 0  0 34]
 [ 0  0  0]]


Below Diagonal Matrix:
 [[ 0  0  0]
 [32  0  0]
 [ 0 58  0]]


Here all the non-diagonal elements are zero.

**Identity Matrix**: A matrix of size nxn where all diagonal values are 1 and other values are 0 is an identity matrix.

4. Create a 4x4 identity matrix with integer numbers.

In [18]:
IM = np.eye(4, dtype=int)

print("4x4 Identity Matrix:\n", IM)

4x4 Identity Matrix:
 [[1 0 0 0]
 [0 1 0 0]
 [0 0 1 0]
 [0 0 0 1]]


**Null Matrix**: A null matrix has all the values as zero.

5. Create a null matrix of integers with shape 6x6.

In [17]:
NM = np.zeros((6,6), dtype=int)
print("6x6 Null Matrix:\n", NM)

6x6 Null Matrix:
 [[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]]


### Matrix Transpose, Inverse, and Determinant

**Transpose**: Transpose operation flips the matrix over its diagonal i.e. replacing rows with coloums and vice-versa.

1. Create a matrix and find its transpose.

In [19]:
# Creating a matrix
OrigMatrix = np.random.randint(low=1, high=101, size=(3, 3))
print("Original Martrix:\n", OrigMatrix)

# Matrix transpose
TransposedMatrix = np.transpose(OrigMatrix)
print("\n\nTransposed Martrix:\n", TransposedMatrix)


Original Martrix:
 [[86 48 83]
 [10 52 38]
 [21 82 52]]


Transposed Martrix:
 [[86 10 21]
 [48 52 82]
 [83 38 52]]


**Inverse**: This operation helps to create a matrix such that when it is multiplied by the original matrix, the result is an identity matrix.

2. Find the inverse of a matrix.

In [27]:
# Create a matrix
OrigMatrix = np.random.randint(low=1, high=101, size=(3, 3))
print("Original Martrix:\n", OrigMatrix)

# Finding the inverse
IM = np.linalg.inv(OrigMatrix)
print("\n\nInverse Martrix:\n", IM)

print("\n\n Multiplicatin of original with Inverse:\n", np.matmul(OrigMatrix, IM))


Original Martrix:
 [[95  5 42]
 [56 75 86]
 [76 77 29]]


Inverse Martrix:
 [[ 0.0097479  -0.00677114  0.00596228]
 [-0.01076718  0.00095791  0.01275315]
 [ 0.00304252  0.01520163 -0.01500435]]


 Multiplicatin of original with Inverse:
 [[ 1.00000000e+00  3.12250226e-17 -5.55111512e-17]
 [-5.03069808e-17  1.00000000e+00  5.55111512e-17]
 [-5.11743425e-17 -2.60208521e-17  1.00000000e+00]]


3. Create 3x3 matrix and find the inverse of it.

In [28]:
# Create a matrix
OrigMatrix = np.random.randint(low=1, high=101, size=(3, 3))
print("Original Martrix:\n", OrigMatrix)

# Finding Inverse 
IM = np.linalg.inv(OrigMatrix)
print("\n\nInverse Martrix:\n", IM)

Original Martrix:
 [[97 84  7]
 [77 48 66]
 [65 39 16]]


Inverse Martrix:
 [[-0.02233186 -0.01324331  0.06439886]
 [ 0.03781331  0.01356481 -0.07249818]
 [-0.00144675  0.02073673 -0.02240605]]


**Determinant**: Determinant of a matrix is a scalar value that helps to characterize the properties of a matrix.

4. Create a matrix of size 3x3 and find the determinant of it.

  To know more about Python libraries, click [here](https://numpy.org/doc/stable/reference/generated/numpy.linalg.det.html)




In [29]:
OrigMatrix = np.random.randint(low=1, high=101, size=(3, 3))
print("Original Martrix:\n", OrigMatrix)

determinant = np.linalg.det(matrix)
print("\n\nDeterminant = ", determinant)

Original Martrix:
 [[55 62 70]
 [86 16 46]
 [82 32 59]]


Determinant =  -0.032212676734701344


5. Create a 4x4 matrix and slice the rows and columns to get the middle sub-matrix and find its determinant.

In [32]:
# Create 4x4 matrix
OrigMatrix = np.random.randint(low=1, high=101, size=(4, 4))
print("Original Martrix:\n", OrigMatrix)


# Slicing of rows and columns
sub_matrix = OrigMatrix[1:3, 1:3]
print("\n\nMiddle Sub-Matrix:\n", sub_matrix)

# Finding the determinant
determinant = np.linalg.det(sub_matrix)
print("\n\nDeterminant = ", determinant)

Original Martrix:
 [[ 4 94 62 97]
 [15 61 12 29]
 [72 49 13 85]
 [10 37 44 28]]


Middle Sub-Matrix:
 [[61 12]
 [49 13]]


Determinant =  205.00000000000006



### Matrix Operations and Functions


### Addition 

1. Create matrix and add a scalar to it.





In [35]:
# Create a matrix
OrigMatrix = np.random.randint(low=1, high=101, size=(3, 3))
print("Original Martrix:\n", OrigMatrix)

Scalar = np.random.randint(low=1, high=101)
print("\n\nScalar = : ", Scalar)

# Scalar Addition
Add_Matrix = OrigMatrix + Scalar
print("\n\nNew Matrix:\n ", Add_Matrix)


Original Martrix:
 [[ 9 38 70]
 [68 60 64]
 [67 99 58]]


Scalar = :  34


New Matrix:
  [[ 43  72 104]
 [102  94  98]
 [101 133  92]]


2. Create two matrices as matrix1 & matrix2. Then, perform addition operation on them.










In [37]:
# Creating matrices
OrigMatrix1 = np.random.randint(low=1, high=101, size=(3, 3))
print("Original Martrix 1:\n", OrigMatrix1)

OrigMatrix2 = np.random.randint(low=1, high=101, size=(3, 3))
print("\n\nOriginal Martrix 2:\n", OrigMatrix2)


# Matrix Addition
SumMatrix = OrigMatrix1 + OrigMatrix2
print("\n\nSum Martrix:\n", SumMatrix)


Original Martrix 1:
 [[69 71 18]
 [ 1 86 52]
 [54  3 61]]


Original Martrix 2:
 [[ 5 37 49]
 [48  3 32]
 [ 8 64 73]]


Sum Martrix:
 [[ 74 108  67]
 [ 49  89  84]
 [ 62  67 134]]


### Subtraction

3. Create a matrix of size 3x6 and subtract a scalar from it.



In [38]:
# Creating matrices
OrigMatrix = np.random.randint(low=1, high=101, size=(3, 6))
print("Original Martrix:\n", OrigMatrix)

Scalar = np.random.randint(low=1, high=101)
print("\n\nScalar = : ", Scalar)

# Scalar Substraction
Sub_Matrix = OrigMatrix - Scalar
print("\n\nNew Matrix:\n ", Sub_Matrix)

Original Martrix:
 [[24 53 46 66 14 54]
 [75 30 53  5 37 54]
 [67 24 74 33 20 72]]


Scalar = :  47


New Matrix:
  [[-23   6  -1  19 -33   7]
 [ 28 -17   6 -42 -10   7]
 [ 20 -23  27 -14 -27  25]]


4. Create two matrices of size 2x3 and subtract second matrix from the first one.

In [39]:
# Creating matrices
OrigMatrix1 = np.random.randint(low=1, high=101, size=(2, 3))
print("Original Martrix 1:\n", OrigMatrix1)

OrigMatrix2 = np.random.randint(low=1, high=101, size=(2, 3))
print("\n\nOriginal Martrix 2:\n", OrigMatrix2)


# Matrix Subtraction
Sub_Matrix = OrigMatrix1 - OrigMatrix2
print("\n\nSubrection Martrix:\n", Sub_Matrix)

Original Martrix 1:
 [[ 59  56  41]
 [ 87 100  68]]


Original Martrix 2:
 [[29 74 54]
 [37 30 21]]


Subrection Martrix:
 [[ 30 -18 -13]
 [ 50  70  47]]


### Division

5. Create a 4x3 matrix and divide it by a scalar.






In [40]:
# Creating matrix
OrigMatrix = np.random.randint(low=1, high=101, size=(4, 3))
print("Original Martrix:\n", OrigMatrix)

Scalar = np.random.randint(low=1, high=101)
print("\n\nScalar = : ", Scalar)

# Scalar Division
Div_Matrix = OrigMatrix / Scalar
print("\n\nNew Matrix:\n ", Div_Matrix)

Original Martrix:
 [[60  7 66]
 [87 12 78]
 [50 82 91]
 [84 48 32]]


Scalar = :  91


New Matrix:
  [[0.65934066 0.07692308 0.72527473]
 [0.95604396 0.13186813 0.85714286]
 [0.54945055 0.9010989  1.        ]
 [0.92307692 0.52747253 0.35164835]]


6. Create two matrices and divide the first matrix by the second one.

In [41]:
# Creating matrices
OrigMatrix1 = np.random.randint(low=1, high=101, size=(2, 2))
print("Original Martrix 1:\n", OrigMatrix1)

OrigMatrix2 = np.random.randint(low=1, high=101, size=(2, 2))
print("\n\nOriginal Martrix 2:\n", OrigMatrix2)

# Matrix Division
Div_Matrix = np.divide(OrigMatrix1, OrigMatrix2)
print("\n\nDivision Martrix:\n", Div_Matrix)



Original Martrix 1:
 [[62 44]
 [95 92]]


Original Martrix 2:
 [[43 89]
 [45 20]]


Division Martrix:
 [[1.44186047 0.49438202]
 [2.11111111 4.6       ]]


### Multiplication

7. Create a matrix and mulitply it by a Scalar.




In [42]:
# Creating matrix
OrigMatrix = np.random.randint(low=1, high=101, size=(4, 3))
print("Original Martrix:\n", OrigMatrix)

Scalar = np.random.randint(low=1, high=101)
print("\n\nScalar = : ", Scalar)


# Scalar Product
Mult_Matrix = OrigMatrix * Scalar
print("\n\nNew Matrix:\n ", Mult_Matrix)

Original Martrix:
 [[25 97 51]
 [54 98 71]
 [50 55 29]
 [23 45 94]]


Scalar = :  49


New Matrix:
  [[1225 4753 2499]
 [2646 4802 3479]
 [2450 2695 1421]
 [1127 2205 4606]]


Multiplication of two matrices is shown in the below diagram:



$A.B = \begin{bmatrix}
a & b\\
c & d\\
\end{bmatrix} \begin{bmatrix}
e & f\\
g & h\\
\end{bmatrix} = \begin{bmatrix}
ae+ bg & af+bh\\
ce+dg & cf+dh\\
\end{bmatrix}$

8. Create two matrices and multiply them.

  Note: Here, the corresponding elements are not multiplied. The multiplication rules are different from addition, subtraction, and division. 

In [43]:
# Creating matrices
OrigMatrix1 = np.random.randint(low=1, high=10, size=(2, 2))
print("Original Martrix 1:\n", OrigMatrix1)

OrigMatrix2 = np.random.randint(low=1, high=10, size=(2, 2))
print("\n\nOriginal Martrix 2:\n", OrigMatrix2)

# Dot Product
mult_mat = np.matmul(OrigMatrix1, OrigMatrix2)
print("\n\nResult Martrix 2:\n", mult_mat)

Original Martrix 1:
 [[5 7]
 [9 2]]


Original Martrix 2:
 [[6 6]
 [7 2]]


Result Martrix 2:
 [[79 44]
 [68 58]]



### Cross Product & Dot Product of Vector and Matrices


*   **Cross Product** : In three-dimensional space, the cross product is a binary operation between two vectors. It produces a perpendicular vector to both vectors.

<center><img src="https://cdn.iisc.talentsprint.com/DLFA/Experiment_related_data/cross_product.png" alt="Drawing" height="300" width="280"/></center>

*   **Dot Product** : The dot product is defined algebraically as the sum of the products of the corresponding entries of the two number sequences. Moreover, it is the product of the Euclidean magnitudes and the cosine of the angle between them.

<br>
<center>
<img src="https://cdn.iisc.talentsprint.com/DLFA/Experiment_related_data/dot_product-01.png" alt="Drawing" height="300" width="300"/></center>


1. Create two matrices and perform cross product on them.

In [44]:
# Creating matrices
v1 = np.array([1, 2, 3])
v2 = np.array([4, 5, 6])

# Cross Product
cp = np.cross(v1, v2)
print("Cross Product = ", cp)

Cross Product =  [-3  6 -3]


2. Apply dot product on the below given vectors:

    i). x = sin 45 + 6i
    
    ii).   y = 42 + 7i



In [46]:
# Taking two 1D arrays
y = [42,7]
x = [sin(np.pi/4),6]

# Calculating dot product using dot() method
dp = np.dot(x, y)
print("Dot Product = ", dp)

Dot Product =  71.698484809835


3. Create two vectors and perform cross product operation on them.


In [47]:
# Creating matrices
v1 = np.array([1, 2, 3])
v2 = np.array([4, 5, 6])

# Cross Product
cp = np.cross(v1, v2)
print("Cross Product = ", cp)

Cross Product =  [-3  6 -3]


### System of Linear Equations

1. Solve the following linear equation for x and y using matrix solution.

  i). 7x + 3y = 34

  ii). 8x + 9y = 50

    Hint: Use the equation for AX = B to solve the equation where A & B are the matrices for values of LHS and RHS of the equation and X consist of x & y variables.
    
    To know more about Python libraries, [here ](https://numpy.org/doc/stable/reference/generated/numpy.linalg.inv.html)
 







Steps to solve:

1. Write the equation in the form of $AX = B$
2. Then, you can write it as -> $X = A^{-1}.B$ 

In [52]:
# Creating matrices
A = [[7,3],[8,9]]
B = [34,50]

# Inverse of matrix A
INV_A = np.linalg.inv(A)

# Dot product of matrix A1 and B
Result = np.matmul(INV_A, B)
print(Result)

print("\n\nx = ", Result[0])
print("y = ", Result[1])

[4. 2.]


x =  3.9999999999999996
y =  2.0000000000000004


2. Find the values of x and y from the system of linear equations given below and verify the solution:

  i). 6x + 5y = 19

  ii). 7x + 9y = 32







Steps to solve:

1. Create two arrays. 

2. Solve for x and y.

3. Compute the dot product of the matrix equation(array) and the solution

4. Use allclose() method to check whether the matrices are element-wise equal or not.

In [54]:
# Creating arrays
A = np.array([[6, 5], [7, 9]])
B = np.array([19, 32])

# Solve a linear matrix equation
X = np.linalg.solve(A, B)

# Verify the solution
np.allclose(np.dot(A,X), B)
# Creating matrices

True

3. Below are pair of three linear equations. Find the values of x, y, & z.

  i). 8x + 3y + 5z = 29

  ii). 4x + 7y + 12z = 54

  iii). 9x + 5y + 11z = 52







In [58]:
# Creating matrices
A = [[8,3,5], [4,7,12], [9,5,11]]
B = [29,54,52]

# Inverse of matrix A
INV_A = np.linalg.inv(A)

# Dot product of A1 and B
X = np.dot(INV_A, B)
print("Result Matrix:", X)

#Verifying Result
np.allclose(np.dot(A,X), B)


Result Matrix: [1. 2. 3.]


True

4. Suppose an automobile company sells 25 bikes and 32 scooties in January. In the next month the sale goes up to 30 bikes and 41 scooties. The total revenue generated is Rs. 506,000 and Rs.620,000 for the respective months. 
Find the cost of bike and scooty using matrix solution.

  Hint: Form the pair of linear equations using the below problem statement and solve the equations to get the answer.

    Assume the cost of bike and scooty as x and y. Moreover, we can create the equation:

       i)   25x + 32y = 506000

       ii)  30x + 41y = 628000



In [60]:
# Create matrices using the values from the equation
A = [[25, 32], [30, 41]]
B = [506000, 628000]

# Inverse of matrix A 
INV_A = np.linalg.inv(A)

# Dot product of A1 and B
X = np.dot(INV_A, B)
print("Result Matrix:", X)

#Verifying Result
np.allclose(np.dot(A,X), B)

print("\nCost of Bike = ", round(X[0]))
print("Cost of Scooty = ", round(X[1]))

Result Matrix: [10000.  8000.]

Cost of Bike =  10000
Cost of Scooty =  8000


### Checking the Linear Dependence and Independence 

We use linear dependence and independence to check the relation of the variables in the real-world dataset.

1. Check linear dependence for the following equations.

  i). 6x + 3y + 5z = 19 

  ii). 2x + y + 7z = 17  

  iii). 16x + 8y + 40z = 104  

  Hint : If the determinant is 0 then it is linearly dependent.





In [ ]:
# Creating arrays
# YOUR CODE HERE


# Finding determinant
# YOUR CODE HERE

# Applying condition
# YOUR CODE HERE

2. Find whether the below linear equations are linearly dependent or independent.

  i). 4x + 2y = 6

  ii). y = -x-2


In [ ]:
a = np.array([[4,2],[1,1]])
print(a)

# Finding determinant
# YOUR CODE HERE

# Applying the condition
# YOUR CODE HERE

### Please answer the questions below to complete the experiment:




Consider the statements given below and answer Q1.


A. Every skew-symmetric matrix of odd order is non-singular

B. If determinant of a square matrix is non-zero, then it is non singular

C. Adjoint of symmetric matrix is symmetric

D. Adjoint of a diagonal matrix is diagonal

In [ ]:
#@title Q.1. Which of the above statements is/are False?
Answer1 = "" #@param ["","Only A","Both A & B","Only B", "Only C", "Both B & C"]


Consider the matrix A and the values of a, b, and c given below for answering Q.2.

$A = \begin{bmatrix}
0 & 2b & c   \\
a & b & -c   \\
a & -b & c   \\
\end{bmatrix}$

<br>

A). a = $\pm \frac{1}{\sqrt{6}} $, b = $\pm \frac{1}{\sqrt{2}}$ , c = $\pm \frac{1}{\sqrt{2}} $


<br>

B). a = $\pm \frac{1}{\sqrt{6}} $, b = $\pm \frac{1}{\sqrt{3}}$ , c = $\pm \frac{1}{\sqrt{2}} $


<br>

C). a = $\pm \frac{1}{\sqrt{2}} $, b = $\pm \frac{1}{\sqrt{6}}$ , c = $\pm \frac{1}{\sqrt{3}} $

<br>

D). a = $\pm \frac{1}{\sqrt{2}} $, b = $\pm \frac{1}{\sqrt{2}}$ , c = $\pm \frac{1}{\sqrt{2}} $

In [ ]:
#@title Q.2. If 'A' is an orthogonal matrix, then, what will be the values of a, b, and c. 
Answer2 = "" #@param ["","Only A","Only B","Only C","Only D", "None of the above"]


In [ ]:
#@title How was the experiment? { run: "auto", form-width: "500px", display-mode: "form" }
Complexity = "" #@param ["","Too Simple, I am wasting time", "Good, But Not Challenging for me", "Good and Challenging for me", "Was Tough, but I did it", "Too Difficult for me"]


In [ ]:
#@title If it was too easy, what more would you have liked to be added? If it was very difficult, what would you have liked to have been removed? { run: "auto", display-mode: "form" }
Additional = "" #@param {type:"string"}


In [ ]:
#@title Can you identify the concepts from the lecture which this experiment covered? { run: "auto", vertical-output: true, display-mode: "form" }
Concepts = "" #@param ["","Yes", "No"]


In [ ]:
#@title  Text and image description/explanation and code comments within the experiment: { run: "auto", vertical-output: true, display-mode: "form" }
Comments = "" #@param ["","Very Useful", "Somewhat Useful", "Not Useful", "Didn't use"]


In [ ]:
#@title Mentor Support: { run: "auto", vertical-output: true, display-mode: "form" }
Mentor_support = "" #@param ["","Very Useful", "Somewhat Useful", "Not Useful", "Didn't use"]


In [ ]:
#@title Run this cell to submit your notebook for grading { vertical-output: true }
try:
  if submission_id:
      return_id = submit_notebook()
      if return_id : submission_id = return_id
  else:
      print("Please complete the setup first.")
except NameError:
  print ("Please complete the setup first.")